Framework/API/src/IFunction.cpp

```diff
@@ -850,65 +860,151 @@ class SetValue : public IFunction::AttributeVisitor<> {
+    bool tempb;
+
+    tempb = (m_value == "true" || m_value == "TRUE" || m_value == "1");
+    evaluateValidator(tempb);
+
+    b = (m_value == "true" || m_value == "TRUE" || m_value == "1");
+  }
   /// Apply if vector
   void apply(std::vector<double> &v) const override {
     if (m_value.empty() || m_value == "EMPTY") {
       v.clear();
       return;
     }
     if (m_value.size() > 2) {
-      // check if the value is in barckets (...)
+      // check if the value is in brackets (...)
       if (m_value.front() == '(' && m_value.back() == ')') {
         m_value.erase(0, 1);
         m_value.erase(m_value.size() - 1);
       }
     }
-    Mantid::Kernel::StringTokenizer tokenizer(m_value, ",", Mantid::Kernel::StringTokenizer::TOK_TRIM);
-    v.resize(tokenizer.count());
+    Kernel::StringTokenizer tokenizer(m_value, ",", Kernel::StringTokenizer::TOK_TRIM);
+    size_t newSize = tokenizer.count();
+
+    // if visitor has an associated validator, first populate temp vec and evaluate against validator.
+    if (m_validator != Mantid::Kernel::IValidator_sptr()) {
+      std::vector<double> tempVec(newSize);
+
+      for (size_t i = 0; i < tempVec.size(); ++i) {
+        tempVec[i] = boost::lexical_cast<double>(tokenizer[i]);
+      }
+      evaluateValidator(tempVec);
+    }
+
+    v.resize(newSize);
     for (size_t i = 0; i < v.size(); ++i) {
       v[i] = boost::lexical_cast<double>(tokenizer[i]);
     }
   }
 
+  /// Evaluates the validator associated with this attribute with regards to input value. Returns error as a string.
+  template <typename T> void evaluateValidator(T &inputData) const {
+    std::string error;
+
+    if (m_validator != Kernel::IValidator_sptr()) {
+      error = m_validator->isValid(inputData);
+    }
+
+    if (error != "") {
+      throw IFunction::validationException("Attribute Set Error: " + error);
+    }
+  }
+
 private:
   mutable std::string m_value; ///< the value as a string
+  mutable Kernel::IValidator_sptr m_validator;
 };
 } // namespace
 
 /** Set value from a string. Throws exception if the string has wrong format
  * @param str :: String representation of the new value
  */
 void IFunction::Attribute::fromString(const std::string &str) {
-  SetValue tmp(str);
+  SetValue tmp(str, m_validator);
   apply(tmp);
 }
 
+/** Set Attribute Name
+ * @param name :: string of attribute name (set via declareAttribute)
+ */
+void IFunction::Attribute::setName(const std::string &name) const { m_name = name; }
+
+/** Set validator to enforce limits on attribute value
+ * @param validator :: shared ptr to validator object
+ */
+void IFunction::Attribute::setValidator(const Kernel::IValidator_sptr &validator) const { m_validator = validator; }
+
+/**
+ *  Evaluates the validator associated with this attribute.
+ */
+void IFunction::Attribute::evaluateValidator() const {
+  std::string dataTypeName;
+  std::string error;
+
+  dataTypeName = type();
+
+  if (dataTypeName == "int") {
```

Comment:
Use `boost::apply_visitor` directly, instead of hand-rolling our own variant of it, see half way down [these docs](https://theboostcpplibraries.com/boost.variant) for an example.
We also have something similar in
https://github.com/mantidproject/mantid/blob/main/qt/scientific_interfaces/ISISReflectometry/GUI/Batch/RowProcessingAlgorithm.cpp#L154

That way this is checked at compile-time and we don't have to faff with dispatch when boost can do it for us

Reply:
This has been implemented, good steer.